### PR TITLE
Enhanced block_from_array with label names

### LINKED ISSUE #685: Allow to specify labels names in `block_from_array`
It would be nice to be able to specify the names used by the labels in `block_from_array`, something like

```py
metatensor.block_from_array(
    array.reshape(1, 3, 3, 1),
    samples="sample_name",
    components=["xyz_1", "xyz_2"],
    properties="properties_name",
)
```

`samples`/`components`/`properties` would be optional and default to the same values as they currently do.

### PR DESCRIPTION
**Pull request by ceriottm**
Fixes #685, adding the ability to create tensor blocks from dense tensors of arbitrary order, specifying the names of each axis. 

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/metatensor/metatensor/actions/artifacts/2045702110.zip)

### CODE REVIEW

**Comment by ceriottm** [comment]
@Luthaf  if this looks reasonable, I'll update the tests.

File: python/metatensor-operations/metatensor/operations/_dispatch.py [hunk 1]
```diff
@@ -527,6 +527,35 @@ def int_array_like(int_list: List[int], like):
...
+    """
+    Creates a tensor of shape ``shape`` filled with ints that
+    enumerate the indices of the entries in the array, e.g.
+    ``shape = (3,2)`` returns ``[[0,0],[0,1],[1,0],[1,1],[2,0],[2,1]]``.
+
+    If the backend is torch and the device is "meta",
+    the device is set to "cpu". This is useful in case where
+    we create labels for a block that is on the meta device.
+    In that case, `int_list` are the labels, and `like` are the block
+    values.
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/_dispatch.py:540]: I'm not sure I like special cases like this inside `_dispatch`. IMO this should be handled at the caller level

File: python/metatensor-operations/metatensor/operations/block_from_array.py [hunk 2]
```diff
@@ -26,15 +33,28 @@ def block_from_array(array) -> TensorBlock:
...
+        The first `d_samples` dimensions in the array will be interpreted as
+        enumerating samples. ``None`` implies one axis named ``"sample"``.
+    :param property_names: A list containing `d_properties` optional property axes
+        names. The last `d_properties` dimensions in the array will be interpreted
+        as enumerating properties. ``None`` implies one axis named ``"property"``.
+    :param component_names: A list containing `d_component` optional component axes
+        names. The middle `d_components` dimensions in the array will be interpreted
+        as enumerating components. ``None`` implies that all the middle dimensions
+        will be considered components, and named ``"component_xxx"``.
+
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/block_from_array.py:46]:
```suggestion
    :param sample_names: A list containing ``d_samples`` sample axes names. The first
        ``d_samples`` dimensions in the array will be interpreted as enumerating
        samples. ``None`` implies a single axis named ``"sample"``.
    :param property_names: A list containing `n_properties` property axes names. The
        last ``n_properties`` dimensions in the array will be interpreted as enumerating
        properties. ``None`` implies a single axis named ``"property"``.
    :param component_names: A list containing ``n_component`` component axes names. The
        middle ``n_components`` dimensions in the array will be interpreted as
        enumerating components. ``None`` implies that all the middle dimensions will be
        considered components, and named ``"component_xxx"``.
```

rst uses double backtick for "code" formatting

File: python/metatensor-operations/tests/block_from_array.py [hunk 3]
```diff
@@ -60,6 +60,40 @@ def test_with_components():
...
+        np.prod(
+            array.shape[: len(sample_names)],
+        )
+        + tuple(
+            array.shape[i]
+            for i in range(len(sample_names), len(array.shape) - len(property_names))
+        )
+        + (len(property_names),)
+    )
+    assert (s1 == s2 for (s1, s2) in zip(block.values.shape, expected_shape))
```
Comment (Luthaf) [inline python/metatensor-operations/tests/block_from_array.py:86]: This works? I would have gone for 

```suggestion
    for actual_shape, expected in zip(block.values.shape, expected_shape)
        assert actual_shape == expected
```

**Review by Luthaf — COMMENTED** [review]
This looks good!

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 1 above)
Comment (ceriottm) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:540]: This was just copied from int_array_like... not clear why it was there, happy to take it out

Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:540]: ah, ok =). I'll see about this one then

File: python/metatensor-operations/metatensor/operations/block_from_array.py [hunk 4]
```diff
@@ -73,31 +96,67 @@ def block_from_array(array) -> TensorBlock:
...
     samples = samples.to(device)
     components = [component.to(device) for component in components]
     properties = properties.to(device)
+    block_shape = [len(samples)]
+    for i in range(d_samples, d_samples + d_components):
+        block_shape.append(shape[i])
+    block_shape.append(len(properties))
+    # reshape array if sample of property axes need to be merged
+    if len(block_shape) != len(shape):
+        array = array.reshape(block_shape)
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/block_from_array.py:160]: Sorry, I missed this earlier. I would rather not reshape here. Instead the user can reshape before calling the function if they want to. If you agree, I can make the change (I found this while cleaning up the code locally)

**Pull request merged**